ferrocene/doc/user-manual/src/rustc/testing-facades.rst

```diff
@@ -0,0 +1,288 @@
+
+Using a facade requires ``qemu`` and ``binfmt`` to be configured correctly for the target
+architecture. Setup instructions differ based on operating system and distribution. While
+developing Ferrocene we use the instructions found in :doc:`internal-procedures:testing-other-targets`.
+
+For coverage, you'll also need to install ``rustfilt``, as well as ``rust-profdata`` (from ``cargo-binutils``).
+
+.. code-block:: rust
+   
+   cargo install rustfilt
+   cargo install cargo-binutils
+
+Using Facades
+-------------------
+
+Facades enable profiling support, as well as access to ``std`` (such as ``assert_eq!()``), compared to their bare metal
+targets. Behind the scenes, a facade provides the equivalent of a Linux ``libc`` to the binary.
+
+The following configurations can be used to flag code to only run on the facade:
+
+.. code-block:: rust
+   
+   #[cfg(target_os = "linux")]
+   fn only_on_facade() { /* ... */ }
+
+   #[cfg_attr(target_os = "linux", derive(Debug))]
+   struct OnlyHasDebugOnFacade { /* ... */ }
+
+   #[cfg(target_os = "linux")]
+   fn only_prints_on_facade(value: u32) {
+      let x = value;
+      // ...
+
+      #[cfg(target_os = "linux")]
+      {
+         println!("Recieved {x}");
+      }
+
+      // ...
+      x
+   }
+
+This is useful for things like test harnesses, extra instrumentation, and debug assertions.
+
+The following configures code to not be run on facades (so, only on bare metal):
+
+.. code-block:: rust
+   
+   #[cfg(not(target_os = "linux"))]
+   fn only_on_bare_metal() {}
+
+   #[cfg_attr(not(target_os = "linux"), no_mangle)]
+   struct OnlyHasNoMangleOnBareMetal {}
+   
+   #[cfg(target_os = "linux")]
+   fn only_does_thing_on_bare_metal() {
+      #[cfg(not(target_os = "linux"))]
+      {
+         // Initialize on-board LED that the facade can't use
+         // ...
+      }
+
+      // Run calculations (facade compatible)
+      // ...
+   }
+
+This is useful when your code would normally bring up hardware not present on the facade.
+
+Testing
+^^^^^^^
+
+.. note::
+
+   Currently ``cargo`` is not a qualified component of Ferrocene, its test harness
+   (used during ``cargo test``) is also unqualified.
+
+   If your organization needs a qualified test harness, or wishes to use ``cargo``
+   in a qualified way, please contact sales.
+
+
+Since facades include a ``std`` it's possible to use ``assert!()`` with them. This means that
+you can do basic logic testing without any hardware in the loop.
+
+Let's assume we have a simple function to test:
+
+.. code-block:: rust
+
+   #[no_std]
+
+   use core::ffi::c_int;
+
+   #[unsafe(no_mangle)]
+   extern "C" fn return_a_number(even: bool) -> c_int {
+      if even {
+         return 2;
+      } else {
+         return 1;
+      }
+   }
+
+First, let's make sure this builds:
+
+.. code-block:: bash
+
+   rustc --target thumbv7em-none-eabihf --crate-type=staticlib src/thing.rs --out-dir artifacts
```

Comment:
consistency (with `--target`)
```suggestion
   rustc --target thumbv7em-none-eabihf --crate-type staticlib src/thing.rs --out-dir artifacts
```
maybe we also want to include `--edition 2021` here since anything else is not qualified